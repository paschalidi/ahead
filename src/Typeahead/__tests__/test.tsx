import React from "react";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { Typeahead } from "../index";

const setup = () => {
  const utils = render(<Typeahead />);
  const typeahead = utils.getByTestId("typeahead");
  const inputElement = utils.getByLabelText("typeahead-input");

  return {
    typeahead,
    inputElement,
    ...utils,
  };
};

test("you should see the rendered input", () => {
  // given
  const { typeahead } = setup();

  // then
  expect(typeahead).toBeInTheDocument();
  expect(typeahead).toMatchInlineSnapshot(`
    <div
      data-testid="typeahead"
    >
      <input
        aria-label="typeahead-input"
        class="sc-bdfBwQ juMTdC"
        placeholder="Type..."
        type="text"
        value=""
      />
    </div>
  `);
});

test(`you should see the suggestion you choose by clicking on the input`, () => {
  // given
  const { inputElement, typeahead, getByLabelText } = setup();

  // when
  fireEvent.change(inputElement, { target: { value: "a" } });
  // and
  fireEvent(
    getByLabelText("Açaí"),
    new MouseEvent("click", {
      bubbles: true,
      cancelable: true,
    })
  );

  // then
  expect(inputElement.value).toBe("Açaí");
  expect(typeahead).toMatchInlineSnapshot(`
    <div
      data-testid="typeahead"
    >
      <input
        aria-label="typeahead-input"
        class="sc-bdfBwQ juMTdC"
        placeholder="Type..."
        type="text"
        value="Açaí"
      />
    </div>
  `);
});

test(`after you chose a suggestion by clicking you should be able to type again in the input`, () => {
  // given
  const { inputElement, typeahead, getByLabelText } = setup();

  // when
  fireEvent.change(inputElement, { target: { value: "a" } });
  // and
  fireEvent(
    getByLabelText("Açaí"),
    new MouseEvent("click", {
      bubbles: true,
      cancelable: true,
    })
  );
  // and
  fireEvent.change(inputElement, { target: { value: "avocado" } });

  // then
  expect(inputElement.value).toBe("avocado");
  expect(typeahead).toMatchInlineSnapshot(`
    <div
      data-testid="typeahead"
    >
      <input
        aria-label="typeahead-input"
        class="sc-bdfBwQ juMTdC"
        placeholder="Type..."
        type="text"
        value="avocado"
      />
      <ul
        class="sc-gsTCUz MsIIl"
        data-testid="suggestions-list"
      >
        <li>
          <button
            aria-label="Avocado"
            class="sc-dlfnbm cHfzSk"
            type="button"
          >
            <span>
              
              <span
                class="coloredText"
              >
                Avocado
              </span>
            </span>
            <span />
          </button>
        </li>
      </ul>
    </div>
  `);
});

test(`after you chose a suggestion by clicking you should be able to type again _and_ make choose another suggestion`, () => {
  // given
  const { inputElement, typeahead, getByLabelText } = setup();

  // when
  fireEvent.change(inputElement, { target: { value: "a" } });
  // and
  fireEvent(
    getByLabelText("Açaí"),
    new MouseEvent("click", {
      bubbles: true,
      cancelable: true,
    })
  );
  // and
  fireEvent.change(inputElement, { target: { value: "avoca" } });
  // and
  fireEvent(
    getByLabelText("Avocado"),
    new MouseEvent("click", {
      bubbles: true,
      cancelable: true,
    })
  );

  // then
  expect(inputElement.value).toBe("Avocado");
  expect(typeahead).toMatchInlineSnapshot(`
    <div
      data-testid="typeahead"
    >
      <input
        aria-label="typeahead-input"
        class="sc-bdfBwQ juMTdC"
        placeholder="Type..."
        type="text"
        value="Avocado"
      />
    </div>
  `);
});

test(`you should see the suggestion you choose by pressing enter on the input`, () => {
  // given
  const { inputElement, typeahead } = setup();

  // when
  fireEvent.change(inputElement, { target: { value: "a" } });
  // and
  fireEvent.keyDown(inputElement, {
    key: "Enter",
    keyCode: 13,
  });

  // then
  expect(inputElement.value).toBe("Açaí");
  expect(typeahead).toMatchInlineSnapshot(`
    <div
      data-testid="typeahead"
    >
      <input
        aria-label="typeahead-input"
        class="sc-bdfBwQ juMTdC"
        placeholder="Type..."
        type="text"
        value="Açaí"
      />
    </div>
  `);
});

test(`after you chose a suggestion by pressing enter you should be able to type again in the input`, () => {
  // given
  const { inputElement, typeahead } = setup();

  // when
  fireEvent.change(inputElement, { target: { value: "a" } });
  // and
  fireEvent.keyDown(inputElement, {
    key: "Enter",
    keyCode: 13,
  });
  // and
  fireEvent.change(inputElement, { target: { value: "avocado" } });

  // then
  expect(inputElement.value).toBe("avocado");
  expect(typeahead).toMatchInlineSnapshot(`
    <div
      data-testid="typeahead"
    >
      <input
        aria-label="typeahead-input"
        class="sc-bdfBwQ juMTdC"
        placeholder="Type..."
        type="text"
        value="avocado"
      />
      <ul
        class="sc-gsTCUz MsIIl"
        data-testid="suggestions-list"
      >
        <li>
          <button
            aria-label="Avocado"
            class="sc-dlfnbm cHfzSk"
            type="button"
          >
            <span>
              
              <span
                class="coloredText"
              >
                Avocado
              </span>
            </span>
            <span />
          </button>
        </li>
      </ul>
    </div>
  `);
});

test(`after you chose a suggestion by clicking you should be able to type again _and_ make choose another suggestion`, () => {
  // given
  const { inputElement, typeahead } = setup();

  // when
  fireEvent.change(inputElement, { target: { value: "a" } });
  // and
  fireEvent.keyDown(inputElement, {
    key: "Enter",
    keyCode: 13,
  });
  // and
  fireEvent.change(inputElement, { target: { value: "avoca" } });
  // and
  fireEvent.keyDown(inputElement, {
    key: "Enter",
    keyCode: 13,
  });

  // then
  expect(inputElement.value).toBe("Avocado");
  expect(typeahead).toMatchInlineSnapshot(`
    <div
      data-testid="typeahead"
    >
      <input
        aria-label="typeahead-input"
        class="sc-bdfBwQ juMTdC"
        placeholder="Type..."
        type="text"
        value="Avocado"
      />
    </div>
  `);
});

test(`you should be able using the keyboard to reach the bottom of the list and chose a suggestion`, () => {
  // given
  const { inputElement, typeahead } = setup();

  // when
  fireEvent.change(inputElement, { target: { value: "plu" } });
  fireEvent.keyDown(inputElement, {
    key: "ArrowDown",
    keyCode: 40,
  });
  fireEvent.keyDown(inputElement, {
    key: "ArrowDown",
    keyCode: 40,
  });
  fireEvent.keyDown(inputElement, {
    key: "ArrowDown",
    keyCode: 40,
  });
  fireEvent.keyDown(inputElement, {
    key: "Enter",
    keyCode: 13,
  });

  // then
  expect(inputElement.value).toBe("Plumcot");
  expect(typeahead).toMatchInlineSnapshot(`
    <div
      data-testid="typeahead"
    >
      <input
        aria-label="typeahead-input"
        class="sc-bdfBwQ juMTdC"
        placeholder="Type..."
        type="text"
        value="Plumcot"
      />
    </div>
  `);
});

test(`you should be able using the keyboard to reach the bottom _and_ then the top of the list and chose a suggestion`, () => {
  // given
  const { inputElement, typeahead } = setup();

  // when
  fireEvent.change(inputElement, { target: { value: "plu" } });
  fireEvent.keyDown(inputElement, {
    key: "ArrowDown",
    keyCode: 40,
  });
  fireEvent.keyDown(inputElement, {
    key: "ArrowDown",
    keyCode: 40,
  });
  fireEvent.keyDown(inputElement, {
    key: "ArrowDown",
    keyCode: 40,
  });
  fireEvent.keyDown(inputElement, {
    key: "ArrowUp",
    keyCode: 38,
  });
  fireEvent.keyDown(inputElement, {
    key: "ArrowUp",
    keyCode: 38,
  });
  fireEvent.keyDown(inputElement, {
    key: "Enter",
    keyCode: 13,
  });

  // then
  expect(inputElement.value).toBe("Japanese plum");
  expect(typeahead).toMatchInlineSnapshot(`
    <div
      data-testid="typeahead"
    >
      <input
        aria-label="typeahead-input"
        class="sc-bdfBwQ juMTdC"
        placeholder="Type..."
        type="text"
        value="Japanese plum"
      />
    </div>
  `);
});

test(`you should be able to search using the fuzzy search`, () => {
  // given
  const { inputElement, getByTestId } = setup();

  // when
  fireEvent.change(inputElement, { target: { value: "buas" } });
  // then
  expect(getByTestId("suggestions-list")).toHaveTextContent("Buddha's hand");
});

const userInputCases = ["cherr", "plum", "avo", "apple", "aç"];

userInputCases.forEach((userInput) =>
  test(`you should see a list of suggestions when user types ${userInput}`, () => {
    // given
    const { inputElement, typeahead, getByTestId } = setup();

    // when
    fireEvent.change(inputElement, { target: { value: userInput } });

    // then
    const listElement = getByTestId("suggestions-list");
    expect(typeahead).toMatchSnapshot();
    expect(listElement).toBeInTheDocument();
  })
);
