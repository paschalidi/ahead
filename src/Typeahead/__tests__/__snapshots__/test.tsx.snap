// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`you should see a list of suggestions when user types apple 1`] = `
<div
  data-testid="typeahead"
>
  <input
    aria-label="typeahead-input"
    class="sc-bdfBwQ dCriQH"
    placeholder="Type..."
    type="text"
    value="apple"
  />
  <ul
    class="sc-gsTCUz MsIIl"
    data-testid="suggestions-list"
  >
    <li>
      <button
        aria-label="Apple"
        class="sc-dlfnbm cHfzSk"
        type="button"
      >
        <span>
          
          <span
            class="coloredText"
          >
            Apple
          </span>
        </span>
        <span />
      </button>
    </li>
    <li>
      <button
        aria-label="Crab apples"
        class="sc-dlfnbm eUEKhV"
        type="button"
      >
        <span>
          Crab 
          <span
            class="coloredText"
          >
            apple
          </span>
        </span>
        <span>
          s
        </span>
      </button>
    </li>
    <li>
      <button
        aria-label="Pineapple"
        class="sc-dlfnbm eUEKhV"
        type="button"
      >
        <span>
          Pine
          <span
            class="coloredText"
          >
            apple
          </span>
        </span>
        <span />
      </button>
    </li>
    <li>
      <button
        aria-label="Star apple"
        class="sc-dlfnbm eUEKhV"
        type="button"
      >
        <span>
          Star 
          <span
            class="coloredText"
          >
            apple
          </span>
        </span>
        <span />
      </button>
    </li>
  </ul>
</div>
`;

exports[`you should see a list of suggestions when user types avo 1`] = `
<div
  data-testid="typeahead"
>
  <input
    aria-label="typeahead-input"
    class="sc-bdfBwQ dCriQH"
    placeholder="Type..."
    type="text"
    value="avo"
  />
  <ul
    class="sc-gsTCUz MsIIl"
    data-testid="suggestions-list"
  >
    <li>
      <button
        aria-label="Avocado"
        class="sc-dlfnbm cHfzSk"
        type="button"
      >
        <span>
          
          <span
            class="coloredText"
          >
            Avo
          </span>
        </span>
        <span>
          cado
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`you should see a list of suggestions when user types aç 1`] = `
<div
  data-testid="typeahead"
>
  <input
    aria-label="typeahead-input"
    class="sc-bdfBwQ dCriQH"
    placeholder="Type..."
    type="text"
    value="aç"
  />
  <ul
    class="sc-gsTCUz MsIIl"
    data-testid="suggestions-list"
  >
    <li>
      <button
        aria-label="Açaí"
        class="sc-dlfnbm cHfzSk"
        type="button"
      >
        <span>
          
          <span
            class="coloredText"
          >
            Aç
          </span>
        </span>
        <span>
          aí
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`you should see a list of suggestions when user types cherr 1`] = `
<div
  data-testid="typeahead"
>
  <input
    aria-label="typeahead-input"
    class="sc-bdfBwQ dCriQH"
    placeholder="Type..."
    type="text"
    value="cherr"
  />
  <ul
    class="sc-gsTCUz MsIIl"
    data-testid="suggestions-list"
  >
    <li>
      <button
        aria-label="Cherry"
        class="sc-dlfnbm cHfzSk"
        type="button"
      >
        <span>
          
          <span
            class="coloredText"
          >
            Cherr
          </span>
        </span>
        <span>
          y
        </span>
      </button>
    </li>
    <li>
      <button
        aria-label="Surinam cherry"
        class="sc-dlfnbm eUEKhV"
        type="button"
      >
        <span>
          Surinam 
          <span
            class="coloredText"
          >
            cherr
          </span>
        </span>
        <span>
          y
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`you should see a list of suggestions when user types plum 1`] = `
<div
  data-testid="typeahead"
>
  <input
    aria-label="typeahead-input"
    class="sc-bdfBwQ dCriQH"
    placeholder="Type..."
    type="text"
    value="plum"
  />
  <ul
    class="sc-gsTCUz MsIIl"
    data-testid="suggestions-list"
  >
    <li>
      <button
        aria-label="Japanese plum"
        class="sc-dlfnbm cHfzSk"
        type="button"
      >
        <span>
          Japanese 
          <span
            class="coloredText"
          >
            plum
          </span>
        </span>
        <span />
      </button>
    </li>
    <li>
      <button
        aria-label="Plum"
        class="sc-dlfnbm eUEKhV"
        type="button"
      >
        <span>
          
          <span
            class="coloredText"
          >
            Plum
          </span>
        </span>
        <span />
      </button>
    </li>
    <li>
      <button
        aria-label="Plumcot"
        class="sc-dlfnbm eUEKhV"
        type="button"
      >
        <span>
          
          <span
            class="coloredText"
          >
            Plum
          </span>
        </span>
        <span>
          cot
        </span>
      </button>
    </li>
  </ul>
</div>
`;
